### ISSUE TITLE
username:fgets (stdin): Resource temporarily unavailable

### ISSUE THREAD

**Issue by RossBoylan** [HenrikBengtsson/ucsf-vpn#12]
When I try to use the 4.0 master release with OpenConnect 7.08 on Debian stretch it failed to connect; the error seems to be the one in the subject line.  I speculate that openconnect is having trouble reading the user name from ucsf-vpn.

Actually, I just got it to work  with a different key, after finding the one I used was already used up.  I don't know if my attempts used it up, or if that was earlier.  So this report may just indicate that better error reporting for failed logins would be helpful, i.e., a message that the key was rejected.

I still have the pulse client in /usr/local/, although it disappeared from my menu *before* I took in the latest updates.  I mention this because your notes indicate that there is a conflict between openconnect and pulse, which I suppose could be in play.

I am running in a VirtualBox vm with bridged networking. 

I am running the program in place in the git working directory and have a .netrc file with the name and password.  It does have some local changes so that I can enter 7 digit Duo keys (which is what I have).  I'll put those in a pull request.

BTW, I never got the program working completely with the pulse client; there seemed to be some issue communicating with the pulse GUI after it was launched.

I originally ran this from a shell inside emacs.  Thinking that was the problem, I ran in a regular terminal.  It failed in the same way.

Final comment: the username request seems to appear after the initial login and after the 2nd factor has been delivered, which seems quite late in the process.

Here's the log on the terminal with debugging:
```shell
ross@secure-stretch:~/src/ucsf-vpn$ date ; bin/ucsf-vpn start --token prompt --verbose --debug
Tue Aug 14 13:20:25 PDT 2018
[DEBUG] Method: openconnect
[DEBUG] Will prompt user for 2FA token
[DEBUG] call: bin/ucsf-vpn start --token prompt --verbose --debug
[DEBUG] action: start
[DEBUG] VPN server: remote.ucsf.edu
[DEBUG] Realm: 'Dual-Factor Pulse Clients'
[DEBUG] user: 
[DEBUG] pwd=<missing>
[DEBUG] token=<prompt>
[DEBUG] verbose: TRUE
[DEBUG] force: FALSE
[DEBUG] skip: FALSE
[DEBUG] dryrun: FALSE
[DEBUG] extras: 
[DEBUG] method: openconnect
[DEBUG] gui: TRUE
[DEBUG] speed: 1.0
[DEBUG] is_online()
[DEBUG] online: TRUE
[DEBUG] public_ip()
[DEBUG] connection_details()
[DEBUG] Querying https://ipinfo.io/ip for public IP:
[DEBUG] Public connection information: {
  "ip": "66.181.128.6",
  "hostname": "m128-6.dsl.rawbw.com",
  "city": "San Francisco",
  "region": "California",
  "country": "US",
  "loc": "37.7484,-122.4160",
  "postal": "94110",
  "org": "AS7961 Raw Bandwidth Communications, Inc."
}
[DEBUG] is_connected(66.181.128.6)
[DEBUG] connection_details()
[DEBUG] Querying https://ipinfo.io/ip for public IP:
[DEBUG] Public connection information: {
  "ip": "66.181.128.6",
  "hostname": "m128-6.dsl.rawbw.com",
  "city": "San Francisco",
  "region": "California",
  "country": "US",
  "loc": "37.7484,-122.4160",
  "postal": "94110",
  "org": "AS7961 Raw Bandwidth Communications, Inc."
}
[DEBUG] Detected .netrc file: /home/ross/.netrc
[DEBUG] - search: remote.ucsf.edu
[DEBUG] - search pattern: ^[ \t]*machine[ \t]+remote.ucsf.edu([ \t]+|$)
[DEBUG] - found: remote.ucsf.edu
[DEBUG] - user=rdboylan
[DEBUG] - pwd=<hidden>
[DEBUG] PROMPT: Asking user to enter one-time token:
Enter 'push', 'phone', 'sms', a 6 or 7 digit Duo token, or press your YubiKey: [DEBUG] Duo token detected
<valid token>
[DEBUG] - token=<hidden>
[DEBUG] call: bin/ucsf-vpn start --token prompt --verbose --debug
[DEBUG] user: rdboylan
[DEBUG] pwd: <hidden>
[DEBUG] token: <hidden>
[DEBUG] opts:  --juniper https://remote.ucsf.edu/pulse --background --user=rdboylan --passwd-on-stdin
[DEBUG] call: sudo openconnect  --juniper https://remote.ucsf.edu/pulse --background --user=rdboylan --passwd-on-stdin --authgroup="Dual-Factor Pulse Clients"
[sudo] password for ross: 
WARNING: Juniper Network Connect support is experimental.
It will probably be superseded by Junos Pulse support.
GET https://remote.ucsf.edu/pulse
Connected to 128.218.97.10:443
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
Got HTTP response: HTTP/1.1 302 Found
GET https://remote.ucsf.edu/dana-na/auth/url_3/welcome.cgi
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
frmLogin
frmLogin
POST https://remote.ucsf.edu/dana-na/auth/url_3/login.cgi
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
Got HTTP response: HTTP/1.1 302 Moved
GET https://remote.ucsf.edu/dana-na/auth/url_3/welcome.cgi?p=more-cred&id=state_17ec986cd0ead339f4dda809effd0845
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
frmLogin
password#2:
POST https://remote.ucsf.edu/dana-na/auth/url_3/login.cgi
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
Got HTTP response: HTTP/1.1 302 Moved
GET https://remote.ucsf.edu/dana-na/auth/url_3/welcome.cgi?p=failed&auth=2
SSL negotiation with remote.ucsf.edu
Connected to HTTPS on remote.ucsf.edu
frmLogin
username:fgets (stdin): Resource temporarily unavailable

[DEBUG] is_online()
[DEBUG] public_ip()
[DEBUG] connection_details()
[DEBUG] Querying https://ipinfo.io/ip for public IP:
[DEBUG] Public connection information: {
  "ip": "66.181.128.6",
  "hostname": "m128-6.dsl.rawbw.com",
  "city": "San Francisco",
  "region": "California",
  "country": "US",
  "loc": "37.7484,-122.4160",
  "postal": "94110",
  "org": "AS7961 Raw Bandwidth Communications, Inc."
}
[DEBUG] is_connected(66.181.128.6)
[DEBUG] connection_details()
[DEBUG] Querying https://ipinfo.io/ip for public IP:
[DEBUG] Public connection information: {
  "ip": "66.181.128.6",
  "hostname": "m128-6.dsl.rawbw.com",
  "city": "San Francisco",
  "region": "California",
  "country": "US",
  "loc": "37.7484,-122.4160",
  "postal": "94110",
  "org": "AS7961 Raw Bandwidth Communications, Inc."
}
[DEBUG] status(FALSE 66.181.128.6)
[DEBUG] public_hostname()
[DEBUG] connection_details()
[DEBUG] Querying https://ipinfo.io/ip for public IP:
[DEBUG] Public connection information: {
  "ip": "66.181.128.6",
  "hostname": "m128-6.dsl.rawbw.com",
  "city": "San Francisco",
  "region": "California",
  "country": "US",
  "loc": "37.7484,-122.4160",
  "postal": "94110",
  "org": "AS7961 Raw Bandwidth Communications, Inc."
}
[DEBUG] public_ip()
[DEBUG] connection_details()
[DEBUG] Querying https://ipinfo.io/ip for public IP:
[DEBUG] Public connection information: {
  "ip": "66.181.128.6",
  "hostname": "m128-6.dsl.rawbw.com",
  "city": "San Francisco",
  "region": "California",
  "country": "US",
  "loc": "37.7484,-122.4160",
  "postal": "94110",
  "org": "AS7961 Raw Bandwidth Communications, Inc."
}
RESULT: Not connected to the UCSF network [m128-6.dsl.rawbw.com (66.181.128.6)]
```

**Comment by HenrikBengtsson** [comment]
In the end of the day, ucsf-vpn is just a neat wrapper for calling `openconnect`, so play around with:
```sh
sudo openconnect  --juniper https://remote.ucsf.edu/pulse --user=rdboylan --authgroup="Dual-Factor Pulse Clients"
```
to see if you can figure output what might go wrong.

**Comment by HenrikBengtsson** [comment]
Oh... while labeling old issues, I just discovered https://github.com/HenrikBengtsson/ucsf-vpn/issues/1 - the same error message was produced way back when the VPN server was non-responding.

**Comment by RossBoylan** [comment]
I got this message again, and tried entering openconnect directly.  It turns out I was using the wrong token.  When login fails, openconnect reprompts for login credentials (as actually appears in the transcript at the top).  Since ucsf-vpn hasn't stuffed anything more on stdin for openconnect, I believe the errors is from openconnect.  I'm not sure why it doesn't just wait on the input, but I guess it can tell stdin is at "end of file."

openconnect has --passwd-on-stdin as on option that might help (unclear if it affects both password and token), but I don't see anything that will avoid the retry on initial failure.

Absent anything better, it might be good to catch the failure and convert it to the more informative message "login failed".

**Comment by HenrikBengtsson** [comment]
The developer version will now detect this and give a more informative error message, e.g.
```sh
$ ucsf vpn start
Enter 'push', 'phone', 'sms', a 6 or 7 digit token, or press your YubiKey: <seven-digit token>
WARNING: Juniper Network Connect support is experimental.
It will probably be superseded by Junos Pulse support.
password#2:
username:fgets (stdin): Resource temporarily unavailable
ERROR: Failed to connect to VPN server (no running OpenConnect process). Likely reason: 2FA token not accepted [ucsf-vpn 5.1.0-9000, OpenConnect 7.08-3ubuntu0.18.04.1]
```
See also https://github.com/HenrikBengtsson/ucsf-vpn/issues/18#issuecomment-643564422
